#!/bin/bash

##############################################################################
# Options:
#   -a input GTF/GFF formatted annotation file name (optional)
#   -i input SAM file name (required; generated by BWA-MEM)
#   -s sequencing strategy (single or paired)
#   -o output (required)
#   -f FASTA reference genome.
#   -m max spanning distance (default: 500,000)
##############################################################################
while getopts ":a:i:s:o:f:m:" opt; do
  case $opt in
  a) GTF_FILE=$OPTARG ;;
  i) INPUT_SAM_FILE=$OPTARG ;;
  s) STRATEGY=$OPTARG ;;
  o) OUTPUT=$OPTARG ;;
  f) FASTA_FILE=$OPTARG ;;
  m) SPANNING=$OPTARG ;;
  \?)
    echo "Invalid option: -$OPTARG"
    exit 1
    ;;
  :)
    echo "Option -$OPTARG requires an argument."
    exit 2
    ;;
  esac
done

#### Check parameters ####
# Check GTF annotation files
if [ -z "$GTF_FILE" ] || [ ! -f "$GTF_FILE" ]; then
  echo "Annotation file does not exist!"
  exit 3
fi

# Check input files
if [ -z "$INPUT_SAM_FILE" ] || [ ! -f "$INPUT_SAM_FILE" ]; then
  echo "Input file does not exist!"
  exit 4
fi

# Control sequencing strategy "single end" o "paired end"
if [ "$STRATEGY" = "single" ]; then
  PAIRED=false
elif [ "$STRATEGY" = "paired" ]; then
  PAIRED=true
else
  echo "error with select the sequencing strategy"
  exit 5
fi

# Check output
if [ -z "$OUTPUT" ]; then
  echo "Output file must be specified!"
  exit 6
fi

# Check if output directory is writable
if [ ! -w "$(dirname "$OUTPUT")" ]; then
  echo "Output directory is not writable!"
  exit 7
fi

# Check fasta file
if [ -z "$FASTA_FILE" ] || [ ! -f "$FASTA_FILE" ]; then
  echo "FASTA file does not exist!"
  exit 8
fi

# Check max spanning distance (default: 500,000)
if [ -z "$SPANNING" ]; then
  SPANNING=500000
fi

#### circRNA identification ####

if [ $PAIRED = "true" ]; then
  perl /usr/local/bin/CIRI.pl -I "$INPUT_SAM_FILE" -A "$GTF_FILE" -F "$FASTA_FILE" -M "$SPANNING" -o "$OUTPUT" -P
else
  perl /usr/local/bin/CIRI.pl -I "$INPUT_SAM_FILE" -A "$GTF_FILE" -F "$FASTA_FILE" -M "$SPANNING" -o "$OUTPUT" -S
fi

# Check SAM file
if [ ! -f "$OUTPUT" ]; then
  echo "Unable to find CIRI output file!"
  exit 9
fi

chmod 777 "$OUTPUT"
